src/net.cpp

```diff
@@ -1923,80 +1712,74 @@ void CConnman::SocketHandler(CMasternodeSync& mn_sync)
     {
         const NodesSnapshot snap{*this, /* filter = */ CConnman::AllNodes, /* shuffle = */ false};
 
+        const auto timeout = std::chrono::milliseconds(only_poll ? 0 : SELECT_TIMEOUT_MILLISECONDS);
+
         // Check for the readiness of the already connected sockets and the
         // listening sockets in one call ("readiness" as in poll(2) or
         // select(2)). If none are ready, wait for a short while and return
         // empty sets.
-        SocketEvents(snap.Nodes(), recv_set, send_set, error_set, only_poll);
+        events_per_sock = GenerateWaitSockets(snap.Nodes());
+        if (events_per_sock.empty() || !Sock::IWaitMany(socketEventsMode, GetModeFileDescriptor(),
```

Comment:
`events_per_sock.empty()` should prevent `IWaitMany` execution for `select` and `poll` only, this check should be done inside the corresponding handlers

Reply:
Not comfortable doing this, I think the responsibility of ensuring the set is populated for the appropriate mode lies with the caller (more so now that the wait code is moved out of `CConnman`). Instead I've made the `events_per_sock.empty()` check conditional on being in a LT SEM since in an ET SEM, we'd be discarding the set anyways.